src/pruna/algorithms/pruning/torch_structured.py

```diff
@@ -224,154 +250,302 @@ def import_algorithm_packages(self) -> Dict[str, Any]:
-        - ignored_layers (list): Ignored layers during pruning.
+    Tuple[Set[nn.Module], Set[nn.Module]]
+        The boundary and exterior modules.
     """
-    if isinstance(model, Opt):
-        ch_groups, num_heads, ignored_layers = get_opt_params(model, smash_config)
+    if target_module == model:  # If we are pruning the entire model, there is no boundary or exterior
+        return set(), set()
+
+    real_nodes = dg.module2node  # includes all modules, parameters, buffers, even internals like autograd nodes
+    name_of = dg._module2name  # modules and parameter with actual names
+    # We only want the modules and parameters that have actual names
+    real_named = set(real_nodes.keys()) & set(name_of.keys())
+
+    # Get the modules and parameters that are inside the target module
+    enc_inside_prms = set(target_module.parameters())
+    enc_inside_modules = set(target_module.modules())
+    enc_inside = enc_inside_prms | enc_inside_modules  # We want both the parameters and the modules
+
+    interior = real_named & enc_inside  # interior is intersection of entire model and target module
+    exterior = real_named - interior  # exterior is the rest of the modules
+
+    def touches_exterior(node):
+        stack, seen = node.inputs + node.outputs, set()
+        while stack:
+            n = stack.pop()
+            if n in seen:
+                continue
+            seen.add(n)
+            mod = n.module
+            # We don't want the auxiliary nodes like autograd nodes
+            # So we only want the modules that have actual names
+            if mod in real_named:
+                return mod in exterior  # If the input or output is in the exterior, then the module touches the exterior
+            stack.extend(n.inputs)
+            stack.extend(n.outputs)
+        return False
+
+    boundary = {m for m in interior if touches_exterior(real_nodes[m])}
+
+    return boundary, exterior
+
+
+def _get_q_head_attr_value(m: nn.Module) -> Optional[int]:
+    """
+    Get the head count attribute value from the module.
+
+    Parameters
+    ----------
+    m : nn.Module
+        The module to get the head count attribute value from.
 
-    elif isinstance(model, Llama):
-        ch_groups, num_heads, ignored_layers = get_llama_params(model, smash_config)
+    Returns
+    -------
+    Optional[int]
+        The head count attribute value.
+    """
+    for attr in _Q_HEAD_ATTRS:
+        if hasattr(m, attr):
+            return getattr(m, attr)
+        if hasattr(m, "config") and hasattr(m.config, attr):
+            return getattr(m.config, attr)
+    return None
 
-    elif isinstance(model, ViT):
-        ch_groups, num_heads, ignored_layers = get_vit_params(model, smash_config)
 
-    elif isinstance(model, imported_modules["timm"].models.convnext.ConvNeXt):
-        ignored_layers = [model.stem, model.head]
-        ch_groups, num_heads = dict(), dict()
+def _get_kv_head_attr_value(m: nn.Module) -> Optional[int]:
```

Comment:
`_get_kv_head_attr_value` and `_get_q_head_attr_value` are almost identical I wonder if we can generalise the function, wdyt?

something like, not tested though :upside_down_face: 
```
def _get_attr_value(module: nn.Module, attr_names: list[str]):
    for attr in attr_names:
        if hasattr(module, attr):
            return getattr(module, attr)
        if hasattr(module, "config") and hasattr(module.config, attr):
            return getattr(module.config, attr)
    return None


def _get_q_head_attr_value(module: nn.Module):
    return _get_attr_value(module, _Q_HEAD_ATTRS)

def _get_kv_head_attr_value(module: nn.Module):
    return _get_attr_value(module, _KV_HEAD_ATTRS)
```